File: Adhiti_Shaili/Day 7,8/Sort matrix diagonally.cpp
class Solution {
public:
    vector<vector<int>> diagonalSort(vector<vector<int>>& mat) {
        int r=mat.size();
        int c=mat[0].size();
        if(r<=1||c<=1)
        {
            return mat;
        }
        
        for(int i=0;i<r;i++)
        {
            vector<int> v;
            int k=i;
            for(int j=0;j<r-i;j++)
            {
                v.push_back(mat[k++][j]);
            }
            sort(v.begin(),v.end());
            k=i;
            int x=0;
            for(int j=0;j<r-i;j++)
            {
                mat[k++][j]=v[x++];
            }
        }
        
        for(int i=c-r;i<c;i++)
        {
            vector<int> v;
            int k=i;
            for(int j=0;j<c-i;j++)
            {
                v.push_back(mat[j][k++]);
            }
            sort(v.begin(),v.end());
            k=i;
            int x=0;
            for(int j=0;j<c-i;j++)
            {
                mat[j][k++]=v[x++];
            }
        }
        for(int i=1;i<c-r;i++)
        {
            vector<int> v;
            int k=i;
            for(int j=0;j<r;j++)
            {
                v.push_back(mat[j][k++]);
            }
            sort(v.begin(),v.end());
            k=i;
            int x=0;
            for(int j=0;j<r;j++)
            {
                mat[j][k++]=v[x++];
            }
        }
        
        return mat;
        
    }
};

File: sanket-dalvi/Day7-8/sortMatrixDiagonally.py
from typing import List


class Solution:
    def diagonalSort(self, mat: List[List[int]]) -> List[List[int]]:
        if not mat or not mat[0]:
            return mat

        def sortD(iRow, jCol):
            i, j = iRow, jCol
            arr = []
            while i < len(mat) and j < len(mat[0]):
                arr.append(mat[i][j])
                i += 1
                j += 1
            arr.sort()
            i, j = iRow, jCol
            k = 0
            while i < len(mat) and j < len(mat[0]):
                mat[i][j] = arr[k]
                i += 1
                j += 1          
                k += 1
        for row in range(len(mat)):
            sortD(row, 0)
        for col in range(1, len(mat[0])):
            sortD(0, col)
        return mat


# Test
print(Solution().diagonalSort([[3, 3, 1, 1], [2, 2, 1, 2], [1, 1, 1, 2]]))
